cql3/expr/expression.cc

```diff
@@ -516,105 +516,30 @@ value_set intersection(value_set a, value_set b, const abstract_type* type) {
-                // token range.  It is impossible for any fetched row not to match now.
-                return true;
-            },
-            [] (const constant&) -> bool {
-                on_internal_error(expr_logger, "is_satisfied_by: A constant cannot serve as the LHS of a binary expression");
-            },
-            [] (const conjunction&) -> bool {
-                on_internal_error(expr_logger, "is_satisfied_by: a conjunction cannot serve as the LHS of a binary expression");
-            },
-            [] (const binary_operator&) -> bool {
-                on_internal_error(expr_logger, "is_satisfied_by: binary operators cannot be nested");
-            },
-            [] (const unresolved_identifier&) -> bool {
-                on_internal_error(expr_logger, "is_satisfied_by: an unresolved identifier cannot serve as the LHS of a binary expression");
-            },
-            [] (const column_mutation_attribute&) -> bool {
-                on_internal_error(expr_logger, "is_satisified_by: column_mutation_attribute cannot serve as the LHS of a binary expression");
-            },
-            [] (const function_call&) -> bool {
-                on_internal_error(expr_logger, "is_satisified_by: function_call cannot serve as the LHS of a binary expression");
-            },
-            [] (const cast&) -> bool {
-                on_internal_error(expr_logger, "is_satisified_by: cast cannot serve as the LHS of a binary expression");
-            },
-            [] (const field_selection&) -> bool {
-                on_internal_error(expr_logger, "is_satisified_by: field_selection cannot serve as the LHS of a binary expression");
-            },
-            [] (const null&) -> bool {
-                on_internal_error(expr_logger, "is_satisified_by: null cannot serve as the LHS of a binary expression");
-            },
-            [] (const bind_variable&) -> bool {
-                on_internal_error(expr_logger, "is_satisified_by: bind_variable cannot serve as the LHS of a binary expression");
-            },
-            [] (const untyped_constant&) -> bool {
-                on_internal_error(expr_logger, "is_satisified_by: untyped_constant cannot serve as the LHS of a binary expression");
-            },
-            [] (const collection_constructor&) -> bool {
-                on_internal_error(expr_logger, "is_satisified_by: collection_constructor cannot serve as the LHS of a binary expression");
-            },
-            [] (const usertype_constructor&) -> bool {
-                on_internal_error(expr_logger, "is_satisified_by: usertype_constructor cannot serve as the LHS of a binary expression");
-            },
-        }, opr.lhs);
+    if (is<token>(opr.lhs)) {
+        // The RHS value was already used to ensure we fetch only rows in the specified
+        // token range. It is impossible for any fetched row not to match now.
```

Comment:
Note, this won't be true one day. We should compute the token and apply the binary operator to it (and if the query planner converted that subexpression to read_command, remove it from filtering).

oh wait, the following comment says exactly that :facepalm: